batchflow/models/torch/layers/conv_complex.py

```diff
@@ -338,3 +341,84 @@ def __init__(self, inputs=None, kernel_size=(3, 5, 7), stride=1, dilation=1, gro
 
     def forward(self, x):
         return Combine(op=self.combine)([layer(x) for layer in self])
+
+    
+
+class DeformableConv2d(nn.Module):
+    """ Deformable convolution: apply convolution with learnable grid sampling locations and feature amplitudes modulation.
+    Grid offsets and modulation scalars are learned with separate convolution layers.
+    Wang, Wenhai, et al. "`InternImage: Exploring Large-Scale Vision Foundation Models with Deformable Convolutions
+    <https://arxiv.org/abs/2211.05778>`_"
+
+    Parameters
+    ----------
+    offset_groups : int
+        `groups` for offset and modulation convolution layers. Default is 1.
+    """
+    LAYER = nn.Conv2d
+
+    def __init__(self, inputs=None, channels=None, kernel_size=3, stride=1, padding='same',
+                 offset_groups=1, groups=1, dilation=1, bias=False):
+        super().__init__()
+        dims = get_num_dims(inputs)
+        if dims != 2:
+            raise NotImplementedError("DeformableConv2d supports only 2d inputs.")
+
+        in_channels = get_num_channels(inputs)
+        channels = safe_eval(channels, in_channels) if isinstance(channels, str) else channels
+        kernel_size = _pair(kernel_size)
+        stride = _pair(stride)
+        dilation = _pair(dilation)
+
+        # 3x3 part of depthwise-separable
+        self.args = {
+            'in_channels': in_channels, 
+            'out_channels': in_channels,
+            'groups': in_channels,
+            'kernel_size': kernel_size,
+            'stride': stride,
+            'dilation': dilation,
+            'bias': bias,
+        }
+
+        
+        self.args.update(compute_padding(padding=padding, shape=inputs.shape[-dims:], kernel_size=kernel_size,
+                                    dilation=dilation, transposed=False, stride=stride))
+        # Depthwise separable conv
+        self.depthwise_layer = self.LAYER(**self.args)
+        self.pointwise_layer = self.LAYER(in_channels=in_channels, out_channels=channels, kernel_size=(1, 1),
+                                           groups=groups, padding=(0, 0), stride=(1, 1), dilation=(1, 1))
+
+
+        # Offset
+        self.args.update({'out_channels': offset_groups * 2 * kernel_size[0] * kernel_size[1],
```

Comment:
I dislike having the same `dict` updated and reused multiple times, especially when it is an instance attribute and supposed to reflect some properties of a layer. Make copies of parameter dicts for `offset` and `modulator` convs

Reply:
or dont store it in instance at all. `padding`, `stride` and `dilation` can be retrieved from `depthwise_layer` the same way you get its `weight` and `bias`